src/frontend/src/optimizer/plan_node/logical_changed_log.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::expr_visitable::ExprVisitable;
+use super::generic::GenericPlanRef;
+use super::utils::impl_distill_by_unit;
+use super::{
+    gen_filter_and_pushdown, generic, ColPrunable, ColumnPruningContext, ExprRewritable, Logical,
+    PlanBase, PlanTreeNodeUnary, PredicatePushdown, RewriteStreamContext, StreamChangedLog,
+    StreamRowIdGen, ToBatch, ToStream, ToStreamContext,
+};
+use crate::error::ErrorCode::BindError;
+use crate::error::Result;
+use crate::optimizer::property::Distribution;
+use crate::utils::{ColIndexMapping, Condition};
+use crate::PlanRef;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LogicalChangedLog {
+    pub base: PlanBase<Logical>,
+    core: generic::ChangedLog<PlanRef>,
+}
+
+impl LogicalChangedLog {
+    pub fn create(input: PlanRef) -> PlanRef {
+        Self::new(input, true, true).into()
+    }
+
+    pub fn new(input: PlanRef, need_op: bool, need_changed_log_row_id: bool) -> Self {
+        let core = generic::ChangedLog::new(input, need_op, need_changed_log_row_id);
+        Self::with_core(core)
+    }
+
+    pub fn with_core(core: generic::ChangedLog<PlanRef>) -> Self {
+        let base = PlanBase::new_logical_with_core(&core);
+        LogicalChangedLog { base, core }
+    }
+}
+
+impl PlanTreeNodeUnary for LogicalChangedLog {
+    fn input(&self) -> PlanRef {
+        self.core.input.clone()
+    }
+
+    fn clone_with_input(&self, input: PlanRef) -> Self {
+        Self::new(input, self.core.need_op, self.core.need_changed_log_row_id)
+    }
+
+    fn rewrite_with_input(
+        &self,
+        input: PlanRef,
+        input_col_change: ColIndexMapping,
+    ) -> (Self, ColIndexMapping) {
+        let changed_log = Self::new(input, self.core.need_op, true);
```

Comment:
Let's move this logic to `logical_rewrite_for_stream`, because it has done more thing than `rewrite_with_input` should contain.